crates/bevy_asset/src/asset_server.rs

```diff
@@ -633,15 +669,113 @@ mod test {
+                extensions == ["not-a-real-extension"]
+            }
+            _ => false,
+        });
+
+        assert_eq!(asset_server.get_load_state(handle), LoadState::NotLoaded);
+    }
+
+    #[test]
+    fn test_invalid_asset_path() {
+        let asset_server = setup(".");
+
+        let path: AssetPath = "an/invalid/path.png".into();
+        let handle = asset_server.get_handle_untyped(path.get_id());
+
+        let err = futures_lite::future::block_on(asset_server.load_async(path.clone(), true))
+            .unwrap_err();
+        assert!(matches!(err, AssetServerError::AssetIoError(_)));
+
+        assert_eq!(asset_server.get_load_state(handle), LoadState::Failed);
+    }
+
+    #[test]
+    fn test_failing_loader() {
+        let dir = create_dir_and_file("fake.fail");
+        let asset_server = setup(dir.path());
+
+        let path: AssetPath = "fake.fail".into();
+        let handle = asset_server.get_handle_untyped(path.get_id());
+
+        let err = futures_lite::future::block_on(asset_server.load_async(path.clone(), true))
+            .unwrap_err();
+        assert!(matches!(err, AssetServerError::AssetLoaderError(_)));
+
+        assert_eq!(asset_server.get_load_state(handle), LoadState::Failed);
+    }
+
+    #[test]
+    fn test_asset_lifecycle() {
+        let dir = create_dir_and_file("fake.png");
+        let asset_server = setup(dir.path());
+        let mut assets = asset_server.register_asset_type::<PngAsset>();
+
+        let path: AssetPath = "fake.png".into();
+        assert_eq!(
+            LoadState::NotLoaded,
+            asset_server.get_load_state(path.get_id())
+        );
+
+        // load the asset
+        let id =
+            futures_lite::future::block_on(asset_server.load_async(path.clone(), true)).unwrap();
+        let handle = asset_server.get_handle_untyped(id);
+
+        // asset is loading
+        assert_eq!(LoadState::Loading, asset_server.get_load_state(id));
+
+        // mimics one full frame
+        let tick = |server: &AssetServer, assets: &mut Assets<PngAsset>| {
+            free_unused_assets_system_impl(server);
+            server.update_asset_storage(assets);
+        };
+
+        tick(&asset_server, &mut assets);
+        // asset should exist and be loaded at this point
+        assert_eq!(LoadState::Loaded, asset_server.get_load_state(id));
+        assert!(assets.get(&handle).is_some());
+
+        // after dropping the handle, next call to `tick` will prepare the assets for removal.
+        drop(handle);
+        tick(&asset_server, &mut assets);
+        assert_eq!(LoadState::Loaded, asset_server.get_load_state(id));
+        assert!(assets.get(id).is_some());
+
+        // second call to tick will actually remove the asset.
+        tick(&asset_server, &mut assets);
+        assert_eq!(LoadState::Loaded, asset_server.get_load_state(id)); // TODO: should this now show `LoadState::NotLoaded`?
+        assert!(assets.get(id).is_none());
```

Comment:
I'm curious is this is the actual behavior we want??
As by now the asset is removed from the `Assets` collection....

Reply:
Could you document https://github.com/bevyengine/bevy/blob/cf221f9659127427c99d621b76c8085c4860e2ef/crates/bevy_asset/src/info.rs#L43-L48 as part of this PR?

From thinking about those, I think we may need a `LoadState::Unloaded`

Reply:
Cool, I documented the behavior and introduce a new `LoadState`.

Reply:
@mockersf 
Curious is the behavior laid out in the tests is the behavior you'd expect.